src/main/java/com/clevercloud/biscuitpulsar/AuthorizationProviderBiscuit.java

```diff
@@ -51,151 +55,84 @@ public AuthorizationProviderBiscuit(ServiceConfiguration conf, PulsarResources p
     }
 
-    @Override
-    public CompletableFuture<Boolean> canProduceAsync(TopicName topicName, String role, AuthenticationDataSource authenticationData) {
-        if (!role.startsWith("biscuit:")) {
-            return defaultProvider.canProduceAsync(topicName, role, authenticationData);
+    private Either<Exception, Authorizer> authorizerFromBiscuitB64Url(String role) {
+        try {
+            Biscuit sealedBiscuit = Biscuit.from_b64url(role.substring("biscuit:".length()), AuthenticationProviderBiscuit.rootKey);
+            Authorizer authorizer = sealedBiscuit.authorizer();
+            return Either.right(authorizer);
+        } catch (NoSuchAlgorithmException | SignatureException | InvalidKeyException | Error ex) {
+            log.error("Failed to deserialize Biscuit from sealed [{}] due to [{}].", role, ex);
+            return Either.left(ex);
         }
+    }
 
-        Either<Error, Verifier> res = verifierFromBiscuit(role);
-        if (res.isLeft()) {
-            return CompletableFuture.failedFuture(new Exception(res.getLeft().toString()));
-        }
-
-        Verifier verifier = res.get();
-        verifier.set_time();
-        verifier.add_fact(topicFact(topicName)).get();
-        verifier.add_fact(topicOperationFact(topicName, TopicOperation.PRODUCE)).get();
-        verifier.add_rule("right(#authority, $tenant, $namespace, $topic, #produce) <- right(#authority, #admin), topic_operation(#ambient, $tenant, $namespace, $topic, #produce)").get();
-        verifier.add_check("check if right(#authority," + topicFactFragment(topicName) + "," + topicOperationSymbol(TopicOperation.PRODUCE)).get();
-        verifier.allow();
-
-        Either verifierResult = verifier.verify(runLimits);
-        log.debug(verifier.print_world());
+    private CompletableFuture<Boolean> authorize(Callable<CompletableFuture<Boolean>> notBiscuitFallback, Boolean withSuperUserFallback,
+                                                 String action, String role, AuthenticationDataSource authenticationData,
+                                                 Set<String> facts, Set<String> rules, Set<String> checks) {
+        if (!role.startsWith("biscuit:")) {
+            try {
+                return notBiscuitFallback.call();
+            } catch (Exception ex) {
+                log.error("Error during notBiscuitFallback call", ex);
+                return CompletableFuture.failedFuture(ex);
+            }
+        }
+
+        return authorizerFromBiscuitB64Url(role)
+                .fold(CompletableFuture::failedFuture,
+                        authorizer -> {
+                            try {
+                                authorizer.set_time();
+                                authorizer.add_check(adminCheck);
+                                for (String fact : facts) authorizer.add_fact(fact);
+                                for (String rule : rules) authorizer.add_rule(rule);
```

Comment:
if there are a lot of rules, it might make sense to pregenerate an authorizer with all the rules, then use [add_token](https://github.com/CleverCloud/biscuit-java/blob/1696b061a3c0cc42053801348b66ca2ac574aea0/src/main/java/com/clevercloud/biscuit/token/Authorizer.java#L110) to insert the token data (that mainly epends on the amount of facts and rules, it might not be useful here)

Reply:
Got it.